### PR TITLE
Close block settings menu after removing block

### LINKED ISSUE #15313: Menu remains open after actions which moves focus
Menu remains open after actions which moves focus

-   **Severity:** Medium
-   **Affected Populations:**
    -   Low-Vision
    -   Motor Impaired
    -   Cognitively Impaired
-   **Platform(s):**
    -   All / Universal
-   **Components affected:**
    -   Block Options

#### Issue description

When the Unified Toolbar is being used, the Options menu remains open
after triggering an action that causes the keyboard focus to be moved to
the edit area. This results in a situation where the focus position
might be underneath the menu. The actions of duplicating a block,
inserting new content before or after a block, or removing a block, all
have this behaviour.

Obscuring the focus position is potentially confusing for sighted
keyboard users, since it may appear as though no item has focus, or make
it unclear as to where the focus position is. Additionally it is simply
difficult to view the content and moving keyboard focus to and from the
toolbar is cumbersome.

#### Remediation Guidance

Whenever an action in the Options menu causes the focus to move, the
menu should close automatically.

#### Relevant standards

-   [2.1.1 Keyboard
    (Level A)](https://www.w3.org/TR/WCAG20/#keyboard-operation-keyboard-operable)
-   [2.4.7 Focus Visible (Level
    AA)](https://www.w3.org/TR/WCAG20/#navigation-mechanisms-focus-visible)

**Note:** This issue may be a duplicate with other existing accessibility-related bugs in this project. This issue comes from the Gutenberg accessibility audit, performed by [Tenon](https://www.tenon.io) and funded by [WP Campus](https://wpcampus.org/). This issue is GUT-101 in Tenon's report

### LINKED ISSUE #15458: Block Settings Menu in Toolbar trying to render for a non-existing (just removed) block
**Describe the bug**

When removing a block via Toolbar (using the menu with three dots -> selecting Remove Block) then the Block Settings Menu will still try to render again, this will result in either:
a) The preceding block's Toolbar being rendered instead with the Block Settings Menu already open (it shouldn't render open by default), or
b) If there is no block selected (e.g. if the paragraph block is unregistered), then this results in an error:
Screen Shot 2019-05-06 at 2 30 17 PM

**To reproduce**

1. Add an Image block as the only block.
2. Remove block via Toolbar:
Screen Shot 2019-05-06 at 2 31 29 PM
3. See the default Paragraph Toolbar opening with the Settings Menu already open.

Steps to reproduce the console error:
1. Unregister `core/paragraph` block, e.g.
```
wp.domReady( () => {
	wp.blocks.unregisterBlockType( 'core/paragraph' );
} );
```
2. Add a block, e.g. Image block as the only block.
3. Remove the block via Toolbar
4. See console / error display in the editor.

- Potentially related issue: https://github.com/WordPress/gutenberg/issues/13276

**Expected behavior**

1. When removing a block via the Toolbar then the Toolbar should render for the next selected block without the Block Settings Menu being open.
2. When removing the last block of a post via Toolbar and a new block isn't added by default (e.g. if the paragraph block is unregistered and alternative default block not enforced) then an error shouldn't occur.

**Screenshots**
Screen Shot 2019-05-06 at 2 40 33 PM

Screen Shot 2019-05-06 at 3 26 13 PM

Tested on the `master` branch as of today.
Also tested with the released 5.6.1 from official plugin repo -- with that version, a console error occurs but the editor doesn't crash.

### PR DESCRIPTION
**Pull request by swissspidy**
## Description

Fixes #15458 by closing the block settings menu after removing a block through that menu.
Also fixes #15313.

This prevents editor crashes when there are no more blocks left but the block settings menu is still rendered as it's open.

## How has this been tested?

Following the steps outlined in #15458 I did this:

1. Enabled fixed toolbar
2. Create block
3. Click on "More options" to open block settings menu
4. Remove block
5. Verify that the menu is closed and no errors are thrown in the console and the editor is also still functioning.

## Types of changes
Bug fix (non-breaking change which fixes an issue)

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate.

### CODE REVIEW

File: packages/block-editor/src/components/block-toolbar/index.js [hunk 1]
```diff
@@ -33,9 +33,9 @@ function BlockToolbar( { blockClientIds, isValid, mode } ) {
 					<BlockSwitcher clientIds={ blockClientIds } />
 					<BlockControls.Slot />
 					<BlockFormatControls.Slot />
+					<BlockSettingsMenu clientIds={ blockClientIds } />
```
Comment (talldan) [inline packages/block-editor/src/components/block-toolbar/index.js]: This change seems to have a couple of unwanted side-effects:
- After selecting 'Edit as HTML' from the block settings menu, users are unable to switch back using the 'Edit visually' option because of the `mode === 'visual'` logic a few lines above.
- Users are unable to remove blocks that become invalid because of a block validation error (due to the `isValid` bool). This one's a little tricky to reproduce, I did the following:
1. Insert a seperator block and save the post
2. Modify the separator's save function so that it returns a `div` instead of an `hr`
3. Reload the post.
4. Observe the block validation error and the settings menu is missing.

File: packages/block-editor/src/components/block-settings-menu/index.js [hunk 2]
```diff
@@ -108,7 +108,10 @@ export function BlockSettingsMenu( { clientIds, onSelect } ) {
 							{ ! isLocked && (
 								<MenuItem
 									className="editor-block-settings-menu__control block-editor-block-settings-menu__control"
-									onClick={ onRemove }
+									onClick={ () => {
+										onClose();
```
Comment (talldan) [inline packages/block-editor/src/components/block-settings-menu/index.js]: This does resolve the issue, but it makes the behaviour of this menu item inconsistent with the others.

**Review by talldan — COMMENTED** [review]
Thanks for the work on this ❤️. 

I thought I'd review as I'd seen this bug mentioned in a design triage, but then I couldn't find the comments related to the triage on the original issue. It turns out there's a separate issue from the accessibility audit - https://github.com/WordPress/gutenberg/issues/15313

I know it broadens the scope a bit, but it'd be great to address this for each of the menu items, as mentioned on that issue.

File: packages/block-editor/src/components/block-toolbar/index.js (hunk 1 above)
Comment (swissspidy) [reply packages/block-editor/src/components/block-toolbar/index.js]: I see, thanks for spotting. It might actually be unrelated to the actual bug, so I‘ll try to revert this change

**Comment by gziolo** [comment]
> I know it broadens the scope a bit, but it'd be great to address this for each of the menu items, as mentioned on that issue.

I don't think it should block this PR.

File: packages/block-editor/src/components/block-settings-menu/index.js (hunk 2 above)
Comment (gziolo) [reply packages/block-editor/src/components/block-settings-menu/index.js]: I think it's fine as the focus needs to be moved to the different block anyway.

File: packages/block-editor/src/components/block-toolbar/index.js (hunk 1 above)
Comment (swissspidy) [reply packages/block-editor/src/components/block-toolbar/index.js]: Reverted in f5dce474e6e5b79ae299336d9ecca5382f60fb28

**Comment by swissspidy** [comment]
> I know it broadens the scope a bit, but it'd be great to address this for each of the menu items, as mentioned on that issue.

Since it was a rather easy fix, I've made that change now as well.

The only case where the settings menu is not closed is when the action (i.e. "Insert After") is performed via a keyboard shortcut.

**Review by afercia — COMMENTED** [review]
Quickly tested and seems to me there's an issue with the Undo/Redo history. To reproduce:
- add a paragraph block with some text
- remove it via the More menu in the top bar
- click Undo
- at this point, the _default_ block appears: expected the paragraph should be restored
- regardless, the "More" button is rendered in the top bar

Screenshot 2019-05-12 at 13 14 49

- click the "More" button: the editor crashes displaying the "Attempt Recovery" dialog

Screenshot 2019-05-12 at 13 16 41

Repeat the steps above, but this time click Undo twice:
- the deleted paragraph is restored
- the "More" button is rendered in the top bar
- same result: click the "More" button and the editor crashes

**Comment by swissspidy** [comment]
Thanks for testing, @afercia! I can reproduce this issue.

> click the "More" button and the editor crashes

I think the issue here is that `wp.data.select('core/block-editor').getSelectedBlockClientIds()` still returns the previously selected block, even though it has been deleted.

So `getSelectedBlockClientIds()` returns a block, `getBlockOrder()` returns an empty array, and  `getBlock( getSelectedBlockClientIds() )` returns `null`.

**Comment by swissspidy** [comment]
ea264779a6aa1aa4f41f2fc997abdd4969d28311 should fix the issue of the _More_ menu being shown in that case, while not re-introducing the side effect originally mentioned in https://github.com/WordPress/gutenberg/pull/15543#discussion_r282733436.

As for the undo/redo history, the issue seems kinda logical.

For comparison, here's how it works on a fresh post:

1. Click on "Add New" to create a new post. You should see the "Start typing..." placeholder of the default block.
2. Click on placeholder.
3. Notice that placeholder text is removed and that the "undo" button is enabled, meaning that you can undo that action.

With this PR here, the following happens:

1. You write something, e.g. "Hello World"
2. You remove the block via the _More_ menu
3. The block gets removed and he default block gets added [one undo action added]
5. The default block gets focused, i.e. the placeholder is removed [second undo action added]

It works correctly when there are multiple paragraph blocks:

1. Write multiple paragraph blocks
2. Remove the last block via the _More_ menu
3. The previous block gets focused
4. Click undo once to restore the deleted block

---

tl;dr the undo/history works as expected, at least from a technical perspective.

However, as a user it indeed seems odd that it would require two undo actions to get back the previous block. I cannot think of a good solution for that off the top of my head. But it should probably be looked at in a separate PR.

**Comment by afercia** [comment]
> However, as a user it indeed seems odd that it would require two undo actions to get back the previous block. I cannot think of a good solution for that off the top of my head. But it should probably be looked at in a separate PR.

Yep this was also previously noted for a similar issue when pressing Enter at the beginning of a block. Will create a separate issue with details.

**Comment by swissspidy** [comment]
Awesome, thanks!

@talldan Would you mind reviewing this again?

File: packages/block-editor/src/components/block-settings-menu/index.js [hunk 3]
```diff
@@ -70,7 +70,10 @@ export function BlockSettingsMenu( { clientIds, onSelect } ) {
 							{ ! isLocked && canDuplicate && (
 								<MenuItem
 									className="editor-block-settings-menu__control block-editor-block-settings-menu__control"
-									onClick={ onDuplicate }
+									onClick={ () => {
+										onClose();
+										onDuplicate();
+									} }
```
Comment (nerrad) [inline packages/block-editor/src/components/block-settings-menu/index.js]: maybe just use lodash flow here and elsewhere?  eg `flow( onClose, onDuplicate )`

**Comment by aduth** [comment]
> I know it broadens the scope a bit, but it'd be great to address this for each of the menu items, as mentioned on that issue.

This was also discussed in #14754 . In fact, perhaps not here, but I suggest we'd go one step further and handle `onClose` within the implementation of `MenuItem`, so as not to make it the responsibility of the implementer to manage (error-prone). The effort in #14843 was partly intended as a step in this direction.

From https://github.com/WordPress/gutenberg/issues/14754#issuecomment-480268587 :

> Part of the problem is also that selecting the menu items does not close the menu. I'm unsure whether it was done so intentionally, but my reading of the recommendations is that the menu [should be closed](https://www.w3.org/TR/wai-aria-practices/#menu).
> 
> I wonder also if it's something we can improve in the abstractions and relationships between [`Dropdown`](https://github.com/WordPress/gutenberg/tree/master/packages/components/src/dropdown) and [`MenuItem`](https://github.com/WordPress/gutenberg/tree/master/packages/components/src/menu-item) components. There's also [`DropdownMenu`](https://github.com/WordPress/gutenberg/tree/master/packages/components/src/dropdown-menu) which... is neither used by the block settings menu, nor itself uses the `MenuItem` component. At least in how it _is_ implemented, it may serve as a suitable replacement for the block settings menu's use of the more generic `Dropdown` component, with the auto-close behavior built in.

File: packages/block-editor/src/components/block-settings-menu/index.js (hunk 3 above)
Comment (nerrad) [reply packages/block-editor/src/components/block-settings-menu/index.js]: Also, using this pattern for onClick callbacks means every time this component is rendered the `onClick` prop value changes.  So it might be better to avoid doing that (I realize this pattern is used elsewhere in this component).

**Review by nerrad — APPROVED** [review]
Looks okay, I haven't tested and I have a few comments that could be addressed.

**Comment by nerrad** [comment]
>  In fact, perhaps not here, but I suggest we'd go one step further and handle onClose within the implementation of MenuItem, so as not to make it the responsibility of the implementer to manage (error-prone)

I concur.

File: packages/block-editor/src/components/block-settings-menu/index.js (hunk 3 above)
Comment (swissspidy) [reply packages/block-editor/src/components/block-settings-menu/index.js]: I'll rewrite to use `flow()` 👍 

> So it might be better to avoid doing that

What would be the preferred approach in such a case then?

**Comment by swissspidy** [comment]
Just so I understand this correctly:

There is an agreement on that this should be fixed and that this PR does solve the issue. However, the way it is solved should be further improved in a more global fashion (perhaps as part of #14843) as other components might also benefit from this. Is that accurate?

File: packages/block-editor/src/components/block-settings-menu/index.js (hunk 3 above)
Comment (nerrad) [reply packages/block-editor/src/components/block-settings-menu/index.js]: > What would be the preferred approach in such a case then?

I missed this was a functional component so should be okay.

**Comment by miina** [comment]
Tested with the steps that originally caused errors: work well.

**Comment by westonruter** [comment]
Also fixes the problem for me. Before I would get an error thrown when removing the initial block. With this PR, no error is raised.

**Comment by kienstra** [comment]
**Works As Expected**

This works here, and the menu doesn't appear anymore when deleting the only block.

# Before 
before-gb

# After
after-gb

**Comment by aduth** [comment]
> > click the "More" button and the editor crashes
> 
> I think the issue here is that `wp.data.select('core/block-editor').getSelectedBlockClientIds()` still returns the previously selected block, even though it has been deleted.

Related: #12327, #12002

>However, the way it is solved should be further improved in a more global fashion (perhaps as part of #14843) as other components might also benefit from this. Is that accurate?

Yes. Would you mind to create an issue to track this?

**Comment by aduth** [comment]
> It's curious to me that an `isValid` condition makes any meaningful difference. Seems like an incidental workaround.

Aside: If I understand correctly, this "fix" is not related to the original scope of the pull request. It would have simplified review if it had been proposed as its own pull request.

**Comment by aduth** [comment]
> I think the issue here is that `wp.data.select('core/block-editor').getSelectedBlockClientIds()` still returns the previously selected block, even though it has been deleted.
> 
> So `getSelectedBlockClientIds()` returns a block, `getBlockOrder()` returns an empty array, and `getBlock( getSelectedBlockClientIds() )` returns `null`.

In my testing, the error occurs here:

https://github.com/WordPress/gutenberg/blob/8960509d97f72bf4bee68e1513a6c3fd3b0d446e/packages/edit-post/src/components/block-settings-menu/plugin-block-settings-menu-group.js#L15

Because `getBlocksByClientId` generates an array of `[ null ]` since, as you mention, the selected blocks remain in state despite having since been deleted:

https://github.com/WordPress/gutenberg/blob/8960509d97f72bf4bee68e1513a6c3fd3b0d446e/packages/edit-post/src/components/block-settings-menu/plugin-block-settings-menu-group.js#L29

Part of this comes down to: We don't prescribe an expected behavior of `getBlocksByClientId` when it's provided an array of client IDs which don't exist as blocks. Per #12327, the specific circumstances under which this invalid array of client IDs exists is not intended, but the underlying question of "What should `getBlocksByClientId` return?" is valid all the same. As an alternative approach, I think the error could be resolved if it only included members for valid client IDs. However, there are downsides of (a) no guarantee that `getBlocksByClientId( clientIds ).length === clientIds.length` and (b) we set expectations in similar _singular_ selectors like `getBlock` to return a `null` value for an unknown block, so a similar expectation may exist here. If we were to say that the result of `getBlocksByClientId` can be expected to include `null` values, then that should be anticipated in any use of the array to guard against property access on a `null` value (i.e. at the error'ing line above).

The proposed condition of `isValid` works only because the `isBlockValid` selector considers whether a block exists:

https://github.com/WordPress/gutenberg/blob/8960509d97f72bf4bee68e1513a6c3fd3b0d446e/packages/block-editor/src/store/selectors.js#L119-L122

This is a pretty loose guarantee, though. For example, hypothetically I imagine an error could still occur if a multi-selection of blocks included a client ID which did not exist, in the rendering of `BlockSettingsMenu` here:

https://github.com/WordPress/gutenberg/blob/8960509d97f72bf4bee68e1513a6c3fd3b0d446e/packages/block-editor/src/components/block-toolbar/index.js#L20-L27

**Comment by swissspidy** [comment]
> Aside: If I understand correctly, this "fix" is not related to the original scope of the pull request. It would have simplified review if it had been proposed as its own pull request.

I disagree about the "not related to the original scope" part. This PR was intended to fix #15458 (editor crash) and coincidentally also fixes #15313. Without this fix the PR would be incomplete as the editor crashes still in the following scenario:

1. Turn on fixed block toolbar setting
1. Add Columns block.
1. Click on appender to create paragraph block within Columns block
1. In the block toolbar, click on three dots and then "Remove block"

You'll get this:

Screenshot 2019-05-21 at 21 41 15

1. Now, click on three dots
1. Editor crashes
1. Attempt recovery
1. Click on three dots again
1. Editor crashes

---

Thanks for the detailed breakdown after that. Is this something we can address in a separate issue/PR or would you prefer it to be handled here? Since you said:

> Yes. Would you mind to create an issue to track this?

I'm asking because I want to get the workflows right here in this project. At the same time, I'd also love to see this fixed rather sooner than later, as this error occurs frequently in our user testing 🙂

**Comment by aduth** [comment]
The changes here include a different behavior that I don't believe should be expected: In `master`, the block "More Options" button is still available in the block toolbar even if the block is invalid. That is no longer the case in the current state of this branch.

I'm not sure it helps with the issues for "Undo", but I've discovered the underlying cause to the original issues reported in #15458:

1. When a block is removed, we try to [select the previous block](https://github.com/WordPress/gutenberg/blob/5c34f2fe327035784c1728572c9d281c9cb53147/packages/block-editor/src/store/actions.js#L496-L498), but we [don't check](https://github.com/WordPress/gutenberg/blob/5c34f2fe327035784c1728572c9d281c9cb53147/packages/block-editor/src/store/actions.js#L128-L134) that a previous block actually exists, so we end up calling `selectBlock( null )`.
2. The `selectBlock` reducer handler will [happily accept](https://github.com/WordPress/gutenberg/blob/5c34f2fe327035784c1728572c9d281c9cb53147/packages/block-editor/src/store/reducer.js#L749-L762) an invalid `clientId` (`null` in this instance, resulting in `getSelectedBlockClientIds()` being `[ null ]` rather than `[]`). We might not need to worry too much about this if the first point is addressed, but it wouldn't hurt to have some sanity checks here to at minimum accept only strings or at most accept only valid client IDs.

`selectBlock( null )` _could_ make sense as an expression of "clear the currently selected block", but that's not how the reducer is implemented to interpret it. It's the `clearSelectedBlock` action which serves this purpose.

In 2f1a82f2e, I add a condition to only select the previous block if one exists. Could you check to see if the `isValid` check is still necessary at this point?

I have some local changes to include an end-to-end test to verify against this regression, but I see another (unrelated) error which occurs here, which we may want to address separately instead. It is an error caused by `ensureDefaultBlock` when the default block type is unregistered.

To the question of process: I understand the urgency of having the user-facing issue resolved. My concern is that urgency is lost as soon as the symptom is treated. If the underlying issue is left to remain, it will inevitably manifest itself again in the future (or in the present by some yet-to-be-discovered circumstances).

**Comment by swissspidy** [comment]
Thanks a ton @aduth! I can confirm that with this change the `isValid` check is no longer necessary. That means the "More Options" button continues to be available in the block toolbar even if the block is invalid.

**Review by aduth — APPROVED** [review]
Thanks for bearing with me. Looks good to me.

Per my previous comment, I'll follow-up with an end-to-end test case which includes a fix for the unrelated issue with `ensureDefaultBlock`.

**Comment by aduth** [comment]
> Per my previous comment, I'll follow-up with an end-to-end test case which includes a fix for the unrelated issue with `ensureDefaultBlock`.

Pull request at #15786 (currently cherry-picks the commit from this branch).
